### ISSUE TITLE
Empty base_url / secure_base_url in database generates segmentation fault

### ISSUE THREAD

**Issue by peec** [magento/magento2#5056]
Tested in Magento: 2.0.7

This bug can only happen if you manually go into the database and edit `web/unsecure/base_url` to empty and `web/secure/base_url` to empty. The priority is probably not that high to fix, but it's important to document because of a very obscure error: **Segmentation fault**.

I finally figured this out, this relates to manually settings base_url / secure/base_url in the database to empty.

I have manually edited these settings on the core_config_data table because of deployment process with environment variable for the domain. The fix / workaround is basically to set `web/unsecure/base_url={{base_url}}` and `web/secure/base_url="{{unsecure_base_url}}"`.

Basically segmentation fault is happening on setup:install / setup:upgrade if you have configured mangeto like described below:
## Steps to reproduce
1. Install Magento from `2.0.7` tag.
2. Install magento
3. Open phpmyadmin in `core_config_data` table: set `web/unsecure/base_url` and `web/secure/base_url` to empty string.
4. Run `bin/magento setup:upgrade` ( also happens with bin/magento setup:install and other commands ).
## Expected result
1. No segmentation fault.
## Actual result
1. Output of command: Segmentation fault
2. If you set  `web/unsecure/base_url={{base_url}}` and `web/secure/base_url="{{unsecure_base_url}}"`.  and clear the cache, the problem is fixed!

**Comment by orlangur** [comment]
Erh... What?!

Why would somebody enter invalid empty value manually in database?

I don't think this is something what could be even fixed from Magento side, more like a PHP bug and should be isolated and reported there.

Also, I don't think that cases when somebody bypasses validation and achieves undesirable system behavior should be documented anywhere. Segmentation faults are not too rare in PHP ecosystem and there is no need to afraid of them.

**Comment by peec** [comment]
It's alright that empty value might be seen as invalid. Though seg fault should not be the default error message for this. This is a coding error somewhere. Useful error messages is good documentation, segmentation fault is not. 

The fact that automatic deployment's will use direct modifications of configuration inside the config table must be accounted for in some case.

You don't just report this as a PHP bug, the root of the problem must be tracked from the Magento 2 code base, and just then reported as a specific bug in PHP.

**Comment by orlangur** [comment]
Now I see what confused you.

> Though seg fault should not be the default error message for this. This is a coding error somewhere. 

This is not Magento message, this is PHP message (or even a message of PHP runner, like web server or php-fpm). You can easily check with `git grep -i`.

> The fact that automatic deployment's will use direct modifications of configuration inside the config table must be accounted for in some case.

Those deployments must not bypass validation which is present in Admin UI. So, if some value is not possible to save via UI, it must not be set directly.

> the root of the problem must be tracked from the Magento 2 code base

Definitely, in case of some valid scenario it would be good to place a workaround for PHP bug inside Magento code. But, again, scenario you provided is just a one of infinite ways to shoot yourself in the foot.

**Comment by IlnitskiyArtem** [comment]
@peec, thanks for reporting this issue. When i'm trying to reproduce your issue, i receive this. 

` PHP Fatal error:  Maximum function nesting level of '200' reached, aborting! in /var/www/html/example/2.0.7/vendor/magento/module-store/Model/Config/Processor/Placeholder.php on line 112
PHP Stack trace:
PHP   1. {main}() /var/www/html/example/2.0.7/bin/magento:0
PHP   2. Symfony\Component\Console\Application->run() /var/www/html/example/2.0.7/bin/magento:25
`
Can you provide more information or confirm that it's the same result which was reproduced by you?

**Comment by IlnitskiyArtem** [comment]
Hi, @peec. Internal ticket MAGETWO-70988, which tracks this GitHub issue, is in our issue backlog.

**Comment by magento-engcom-team** [comment]
@peec, thank you for your report.
We've created internal ticket(s) MAGETWO-70988 to track progress on the issue.

**Comment by magento-engcom-team** [comment]
Hi @peec 
We are closing the issue now, as intentionally putting invalid values into the database may indeed cause any kind of errors, and it does not look reasonable to take into account every case like this.

Thank you for collaboration.
